python/ray/air/session.py

```diff
@@ -0,0 +1,233 @@
+import abc
+from enum import Enum, auto
+import logging
+from typing import Optional, Dict
+
+from ray.air.checkpoint import Checkpoint
+
+logger = logging.getLogger(__name__)
+
+
+class SessionType(Enum):
+    TRAIN = auto()
+    TUNE = auto()
+
+
+class Session(abc.ABC):
+    """The canonical session interface that both Tune and Train session implements.
+
+    User can interact with this interface to get session information,
+    as well as reporting metrics and saving checkpoint.
+    """
+
+    @property
+    @abc.abstractmethod
+    def type(self) -> SessionType:
+        """The type of the session.
+
+        Returns:
+            The type of the session.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def report(self, metrics: Dict, checkpoint: Optional[Checkpoint] = None) -> None:
+        """Report metrics and optionally save checkpoint.
+
+        The checkpoint is guaranteed to be saved onto Driver node.
+
+        Each invocation of this method will automatically increment the underlying
+        iteration number. The physical meaning of this "iteration" is defined by
+        user (or more specifically the way they call ``report``).
+        It does not necessarily map to one epoch.
+
+        This API is supposed to replace the legacy ``tune.report``,
+        ``with tune.checkpoint_dir``, ``train.report`` and ``train.save_checkpoint``.
+        Please avoid mixing them together.
+
+        There is no requirement on what is the underlying representation of the checkpoint.
+        All forms are accepted and (will be) handled by AIR in an efficient way.
+
+        Example:
+            .. code-block: python
+
+                from ray.air.session import get_session
+                from ray.air.checkpoint import Checkpoint
+                ########## Using it in the *per worker* train loop (TrainSession) ###################
+                def train_func():
+                    session = get_session()
+                    model = build_model()
+                    # ...
+
+                    # TEMPORARY!!
+                    # This is bad if user has to do this. Also we lose track of the ownership of the
+                    # temporary directory and when can clean it up.
+                    # Can be avoided, if we expose session.local_dir for one to write file to upfront.
```

Comment:
Or alternatively, could we change working directory for train workers in a similar fashion as Tune?

Reply:
Train always uses Tune, so doesn't it get chdir automatically already?